kernel/thread_monitor.c

```diff
@@ -83,3 +83,56 @@ void k_thread_foreach_unlocked(k_thread_user_cb_t user_cb, void *user_data)
 	k_spin_unlock(&z_thread_monitor_lock, key);
 
 }
+
+#ifdef CONFIG_SMP
+void k_thread_foreach_my_core(k_thread_user_cb_t user_cb, void *user_data)
+{
+	struct k_thread *thread;
+	k_spinlock_key_t key;
+
+	__ASSERT(user_cb != NULL, "user_cb can not be NULL");
+
+	/*
+	 * Lock is needed to make sure that the _kernel.threads is not being
+	 * modified by the user_cb either directly or indirectly.
+	 * The indirect ways are through calling k_thread_create and
+	 * k_thread_abort from user_cb.
+	 */
+	key = k_spin_lock(&z_thread_monitor_lock);
+
+	SYS_PORT_TRACING_FUNC_ENTER(k_thread, foreach);
+
+	for (thread = _kernel.threads; thread; thread = thread->next_thread) {
+		if (thread->base.cpu == arch_proc_id())
```

Comment:
I have three issues with `if (thread->base.cpu == arch_proc_id())`.

First -- arch_proc_id(). Although this returns a unique identifier per CPU, it is not guaranteed to match the cpu ID numbering that is used within the zephyr kernel. As a result, when thread->base.cpu is set in kernel/include/kswap.h::do_swap() via `new_thread->base.cpu = arch_curr_cpu()->id` the IDs may match on some platforms, but not all.

Second --  Near as I can tell, thread->base.cpu is only set in do_swap() which is called when a thread does a context switch. Unless I have missed it, it does not look like it gets updated when exiting an interrupt. Thus, base.cpu is not currently reliable.

Third -- CPU Masking. base.cpu identifies the last CPU on which the thread executed. If CPU masking is enabled and the set of CPU to which the thread is pinned changed while it was blocked on some kernel object and it has not yet had a chance to run on the new CPU, does it really belong with the threads associated with its old CPU?

Reply:
FYI, the second issue I identified above has been fixed and merged.

Reply:
As you know the Intel ADSP is quite unusual Zephyr platform, and this change quite possibly is only needed for it. So I wonder if we should make the new "current_cpu"-functions optional on some new Kconfig option, and possibly make that option dependent on SOC_FAMILY_INTEL_ADSP?

Then the practical side. I so what should I use here, if not (thread->base.cpu == arch_proc_id())? Would calculating the arch_curr_cpu() offset in struct z_kernel -> cpus[] table be better?